### PR TITLE
add: Broker service MVP with buy operation

### PR DESCRIPTION
**Pull request by rgommezz**

### CODE REVIEW

File: packages/broker/constants/index.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import * as path from "path";
+import * as dotenv from "dotenv";
+
+dotenv.config({
+  path: path.join(__dirname, "..", ".env"),
+}) ;
+
+export const config = {
+  wallet: {
+    secret:  process.env.WALLET_SECRET,
```
Comment (rgommezz) [inline packages/broker/constants/index.ts:10]: Rename this to `decriptionKey`

File: packages/broker/constants/index.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    encryptAlgo: process.env.WALLET_ENCRYPT_ALGO || 'aes-256-cbc'
```
Comment (rgommezz) [inline packages/broker/constants/index.ts:11]: Rename this to `decriptionAlgorithm`

File: packages/price-notification/constants/index.ts [hunk 3]
```diff
@@ -7,8 +7,14 @@ dotenv.config({
 
 export const config = {
   manifest: process.env.MANIFEST_PATH || path.join(__dirname, "..", "configs", "manifest.json"),
+  wallet: {
+    privateKey: process.env.WALLET_PRIVATE_KEY,
+    encryptAlgo: process.env.WALLET_ENCRYPT_ALGO || "aes-256-cbc",
```
Comment (rgommezz) [inline packages/price-notification/constants/index.ts:12]: Rename this to `encryptionAlgorithm`

File: packages/price-notification/constants/index.ts [hunk 4]
```diff
@@ -7,8 +7,14 @@ dotenv.config({
...
+    secret: process.env.WALLET_SECRET
```
Comment (rgommezz) [inline packages/price-notification/constants/index.ts:13]: Rename this to `encryptionSalt`

File: packages/price-notification/index.ts [hunk 5]
```diff
@@ -52,6 +54,15 @@ const runPoll = async (alert: Alert) => {
...
             });
             result.shouldStop = true;
+            // Communicate with broker service
+            await axios.post(`${config.api.broker}` + "/swap", {
+              amount,
+              tokenIn: "USDC",
+              tokenOut: token,
+              network,
+              exchange,
+              walletHash: encrypt(config.wallet.privateKey as string),
```
Comment (rgommezz) [inline packages/price-notification/index.ts:64]: Rename this to `walletPrivateKey`

File: packages/price-notification/utils/crypto.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import crypto from "crypto";
+import { config } from "../constants";
+
+export const encrypt = (text: string) => {
+  const iv = crypto.randomBytes(16);
+
+  const cipher = crypto.createCipheriv(
+    config.wallet.encryptAlgo as string,
```
Comment (rgommezz) [inline packages/price-notification/utils/crypto.ts:8]: Rename this to `encryptionAlgorithm`

File: packages/price-notification/utils/crypto.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    config.wallet.secret as string,
```
Comment (rgommezz) [inline packages/price-notification/utils/crypto.ts:9]: Rename this to `encryptionSalt`

File: packages/broker/controllers/swap.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
+import * as ethers from "ethers";
+import { ContractTransaction } from "ethers";
+import { networks } from "../constants";
+import * as express from "express";
+import safeAsync from "../utils/safeAsync";
+import { decrypt } from "../utils/crypto";
+
+interface Body {
```
Comment (juliantellez) [inline packages/broker/controllers/swap.ts:8]: is this the request? is this express lingo? maybe requestBody? so it tells us what it is

File: packages/broker/controllers/swap.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  amount: number;
+  tokenIn: string;
+  tokenOut: string;
```
Comment (juliantellez) [inline packages/broker/controllers/swap.ts:11]: These two need an comment or explanation, is hard to understand what they do

File: packages/broker/controllers/swap.ts [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+  );
+
+  const TokenAmountIn = ethers.utils.parseUnits(amount.toString(), 18);
+  const amounts = await routerContract.getAmountsOut(TokenAmountIn, [
+    tokenInAddress,
+    tokenOutAddress,
+  ]);
+  const TokenAmountOut = amounts[1];
+
+  // Approval transaction to only allow the protocol to spend the intended amount
```
Comment (juliantellez) [inline packages/broker/controllers/swap.ts:81]: is this approval a one time only transaction?

File: packages/broker/controllers/swap.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+  await tryTransaction(
+    () =>
+      tokenInContract
+        .connect(wallet)
+        .approve(routerContract, TokenAmountIn, { gasLimit: 15000000 }),
```
Comment (juliantellez) [inline packages/broker/controllers/swap.ts:86]: where does this limit come from?

File: packages/price-notification/index.ts [hunk 12]
```diff
@@ -52,6 +54,15 @@ const runPoll = async (alert: Alert) => {
               text: `Price of ${token} is: ${price}. Action: ${action}`,
             });
             result.shouldStop = true;
+            // Communicate with broker service
+            await axios.post(`${config.api.broker}` + "/swap", {
+              amount,
+              tokenIn: "USDC",
```
Comment (juliantellez) [inline packages/price-notification/index.ts:60]: this looks like it needs abstracting, as it is currently an assumption

**Review by juliantellez — APPROVED** [review]
Most comments are about docs and future work... LGTM otherwise

File: packages/broker/controllers/swap.ts (hunk 8 above)
Comment (rgommezz) [reply packages/broker/controllers/swap.ts:8]: Hesho compare

**Review by juliantellez — APPROVED** [review]

**Pull request merged**